### PR TITLE
Package cleanup

### PR DESCRIPTION
**Pull request by gselzer**
Goals of this PR:

* [x] Refactor imagej/ directory into src/imagej/ directory. See [this article](https://hynek.me/articles/testing-packaging/) for the benefits
* [x] Minimize setup.py in favor of setup.cfg. This seems to be the [preferred approach](https://docs.python.org/3/distutils/configfile.html) nowadays.
* [x] Add a linting strategy (using black)
* [x] Add a code coverage strategy
* [x] Allow these features to block merge through Github Actions

### CODE REVIEW

**Comment by gselzer** [comment]
Differences between the old and new distributions:
* We no longer have a `description-file` key in `setup.cfg` - this [SO post](https://stackoverflow.com/questions/60084128/does-description-file-in-setup-cfg-section-metadata-have-any-effect) suggests it is not used by distutils nor setuptools, and we now have `long-description`.

File: .github/workflows/python-test-conda.yml [hunk 1]
```diff
@@ -16,36 +16,107 @@ jobs:
...
       run: |
-        pytest test/ -p no:faulthandler --legacy=false
-    - name: Test Fiji Is Just ImageJ(2)
+        python -m pip install --upgrade pip
+        python -m pip install -e ".[test]"
+
+    - name: Test ImageJ
       run: |
-        pytest test/ -p no:faulthandler --ij=sc.fiji:fiji
+        ./test.sh
```
Comment (ctrueden) [inline .github/workflows/python-test-conda.yml:48]: Yay, I'm glad we can just call this shell script now.

File: src/imagej/config.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+__author__ = "Curtis Rueden, Michael Pinkert, Ed Evans, Yang Liu, Mark Hiner"
+
+__major__ = 1
+__minor__ = 1
+__patch__ = 2
+__tag__ = "dev0"
+__version__ = f"{__major__}.{__minor__}.{__patch__}.{__tag__}".strip(".")
```
Comment (ctrueden) [inline src/imagej/config.py:7]: This is now duplicated here and in `setup.cfg`.

Is there any way for a module to be able to self-report its version in a DRY manner? I'm concerned these two spots may get out of sync.

File: src/pyimagej.egg-info/PKG-INFO [hunk 3]
```diff
@@ -0,0 +1,126 @@
+Metadata-Version: 2.1
```
Comment (ctrueden) [inline src/pyimagej.egg-info/PKG-INFO:1]: I think you committed the `src/pyimagej.egg-info` directory by accident, no? Can you please purge it from the history of this PR?

**Review by ctrueden — CHANGES_REQUESTED** [review]
Looks great! Just two things:

1. A question: should we remove `imagej.config` version and author constants? It's replicated from setup.cfg, and I'm concerned about skew over time. I guess we can live with pyimagej not being able to self-report its version, although I would love to know if there is a Python best practice for this. It seems `setup.cfg` is not bundled with installed packages unfortunately, AFAICT, so reading that is not an option...
2. It looks like `src/pyimagej.egg-info` got committed by accident. This should be in a `.gitignore` as it is autogenerated at packaging time, no?

File: src/pyimagej.egg-info/PKG-INFO (hunk 3 above)
Comment (gselzer) [reply src/pyimagej.egg-info/PKG-INFO]: :vomiting_face:

It shouldn't be part of the commit history anymore, with 616634c

File: .github/workflows/python-test-conda.yml (hunk 1 above)
Comment (gselzer) [reply .github/workflows/python-test-conda.yml:48]: :confetti_ball:

File: src/imagej/config.py (hunk 2 above)
Comment (gselzer) [reply src/imagej/config.py]: Probably. I just removed this in 59406c0.

Let's address versioning in a new issue. See https://github.com/imagej/pyimagej/issues/187

Comment (gselzer) [reply src/imagej/config.py]: Drat, I forgot this was called in `__init__.py`

For now, we'll leave the version construction in `setup.py` :confused:

Hopefully we can devise a better way to do the versioning in that issue.

**Comment by ctrueden** [comment]
@gselzer Looks like the tests (relating to the version logic?) need fixing before we can merge.

**Comment by gselzer** [comment]
> @gselzer Looks like the tests (relating to the version logic?) need fixing before we can merge.

Yup, working on it now!

**Comment by gselzer** [comment]
@ctrueden the versioning is fixed, however I just found another problem. In the tarball, there are no source files!
```
(tmp) : gselzer@gselzer-OptiPlex-5090 ~/code/imagej/pyimagej/dist (master) [tmp] »
tar xzvf pyimagej-1.1.2.dev0.tar.gz                                                                                                                                                                                                                    #  0 {2022-04-13 14:40:56}
pyimagej-1.1.2.dev0/
pyimagej-1.1.2.dev0/PKG-INFO
pyimagej-1.1.2.dev0/test/
pyimagej-1.1.2.dev0/test/test_imagej.py
pyimagej-1.1.2.dev0/README.md
pyimagej-1.1.2.dev0/LICENSE.txt
pyimagej-1.1.2.dev0/setup.cfg
pyimagej-1.1.2.dev0/MANIFEST.in
pyimagej-1.1.2.dev0/pyimagej.egg-info/
pyimagej-1.1.2.dev0/pyimagej.egg-info/PKG-INFO
pyimagej-1.1.2.dev0/pyimagej.egg-info/top_level.txt
pyimagej-1.1.2.dev0/pyimagej.egg-info/dependency_links.txt
pyimagej-1.1.2.dev0/pyimagej.egg-info/requires.txt
pyimagej-1.1.2.dev0/pyimagej.egg-info/entry_points.txt
pyimagej-1.1.2.dev0/pyimagej.egg-info/SOURCES.txt
pyimagej-1.1.2.dev0/setup.py
pyimagej-1.1.2.dev0/imagej/
pyimagej-1.1.2.dev0/imagej/config.py
pyimagej-1.1.2.dev0/imagej/__init__.py
pyimagej-1.1.2.dev0/imagej/stack.py
pyimagej-1.1.2.dev0/imagej/dims.py
```

I'll let you know when I fix it.

**Comment by gselzer** [comment]
> @ctrueden the versioning is fixed, however I just found another problem. In the tarball, there are no source files!

This has been fixed, but something else broke, as now when we write
```
python -m pip install -e ".[test]"
```
the base packages are not being added :confused:

**Comment by gselzer** [comment]
> > @ctrueden the versioning is fixed, however I just found another problem. In the tarball, there are no source files!
> 
> This has been fixed, but something else broke, as now when we write
> 
> ```
> python -m pip install -e ".[test]"
> ```
> 
> the base packages are not being added confused

ALIRIGHT. The problem was [this line](https://github.com/imagej/pyimagej/commit/ecab46e425e7e21d4ca734897b8288b45ec7c0f8#diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52R42). I added it as I thought it was needed to ensure that the `src` directory was included, but that was not needed. The latest commit no longer has this.

I think we are ready to merge @ctrueden. My brain hurts.

**Comment by ctrueden** [comment]
@gselzer Thanks for all your hard work. Reminder to myself: before merging this, do the same check I did with scyjava where I diffed the tarballs from before and after to make sure everything is still good.

**Comment by gselzer** [comment]
> @gselzer Thanks for all your hard work. Reminder to myself: before merging this, do the same check I did with scyjava where I diffed the tarballs from before and after to make sure everything is still good.

Good idea. I checked much of it already, but it would be good to have a second set of eyes on it.

**Review by ctrueden — CHANGES_REQUESTED** [review]
The FIXMEs noted in this WIP commit needs to be finished before this PR can be merged.

And then, the same improvements that were made on this PR need to be made for the other components of our Python stack, including jgo, scyjava, and imglyb.

**Comment by gselzer** [comment]
@ctrueden wrote:
> \# CTR FIXME: Change these rules to prevent the
> \# kitchen sink from being packaged into the sdist.

I don't think we should be excluding anything more. The source distribution looks pretty good to me?
```bash
(scyjava) : gselzer@gselzer-OptiPlex-5090 ~/code/imagej/pyimagej/dist (package-cleanup*) [scyjava] »
ls -R pyimagej-1.1.2.dev0                                                                                                                                                                                                                               #  0 {2022-04-15 8:31:47}
pyimagej-1.1.2.dev0:
conftest.py  doc/  LICENSE.txt  MANIFEST.in  PKG-INFO  pyproject.toml  README.md  setup.cfg  setup.py  src/  test/

pyimagej-1.1.2.dev0/doc:
10-Troubleshooting.ipynb               3-Sending-Data-to-Java.ipynb             6-Working-with-Images.ipynb                 9-Working-with-Large-Images.ipynb  doc-images/  Initialization.md  rtd/                Troubleshooting.md
1-Starting-PyImageJ.ipynb              4-Retrieving-Data-from-Java.ipynb        7-Running-Macros-Scripts-and-Plugins.ipynb  cellprofiler/                      examples/    Install.md         sample-data/        Xvfb.md
2-Opening-and-Displaying-Images.ipynb  5-Convenience-methods-of-PyImageJ.ipynb  8-Discover-and-run-ImageJ-commands.ipynb    Development.md                     Headless.md  README.md          Segmentation.ipynb

pyimagej-1.1.2.dev0/doc/cellprofiler:
figure.svg  MakeBBBC030GT_ManuallyFix.cppipe  README.md  RunImageJMacroWorkflow/  RunImageJScriptRollingBall/  RunImageJScriptSegmentation/

pyimagej-1.1.2.dev0/doc/cellprofiler/RunImageJMacroWorkflow:
__pycache__/  RunImageJBallWithMacro.cppipe  test_ball_macro.py

pyimagej-1.1.2.dev0/doc/cellprofiler/RunImageJMacroWorkflow/__pycache__:
test_ball_macro.cpython-38-pytest-7.0.1.pyc  test_ball_macro.cpython-39-pytest-7.1.1.pyc

pyimagej-1.1.2.dev0/doc/cellprofiler/RunImageJScriptRollingBall:
RunImageJBall.cppipe  test_ball.bsh

pyimagej-1.1.2.dev0/doc/cellprofiler/RunImageJScriptSegmentation:
classifier.model  RunImageJScriptSegmentation_withGT.cppipe  test.bsh

pyimagej-1.1.2.dev0/doc/doc-images:
imagej_ui_macro_recorder.png  macro_recorder_find_maxima.png  test_still_analyze_particles.png  test_still_find_maxima.png

pyimagej-1.1.2.dev0/doc/examples:
blob_detection_interactive.py  blob_detection_xvfb.py  README.md

pyimagej-1.1.2.dev0/doc/rtd:
make.bat  README.md  source/

pyimagej-1.1.2.dev0/doc/rtd/source:
api.rst  conf.py  index.rst  pyimagej.rst  requirements.txt  submodules.rst

pyimagej-1.1.2.dev0/doc/sample-data:
test_image.tif  test_still_stack.tif  test_still.tif  test_timeseries.tif

pyimagej-1.1.2.dev0/src:
imagej/  pyimagej.egg-info/

pyimagej-1.1.2.dev0/src/imagej:
config.py  dims.py  __init__.py  stack.py

pyimagej-1.1.2.dev0/src/pyimagej.egg-info:
dependency_links.txt  entry_points.txt  PKG-INFO  requires.txt  SOURCES.txt  top_level.txt

pyimagej-1.1.2.dev0/test:
test_imagej.py
```

There is a good argument for adding the conda environment files to the sdist, which I just did. We can also add `test.sh`.

This gives the diff:
```bash
(tmp5) : gselzer@gselzer-OptiPlex-5090 ~/code/imagej/pyimagej (package-cleanup*) [tmp5] »
/usr/bin/diff -rq . dist/pyimagej-1.1.2.dev0                                                                                                                                                                                                            #  2 {2022-04-15 9:34:43}
Only in .: codecov.yml
Only in .: dist
Only in ./doc/cellprofiler/RunImageJMacroWorkflow: __pycache__
Only in ./doc/rtd: Makefile
Only in .: .git
Only in .: .github
Only in .: .gitignore
Only in .: .mailmap
Only in dist/pyimagej-1.1.2.dev0: PKG-INFO
Only in .: __pycache__
Only in .: pyimagej.egg-info
Only in .: .pytest_cache
Only in .: .readthedocs.yaml
Files ./setup.cfg and dist/pyimagej-1.1.2.dev0/setup.cfg differ
Only in ./src/imagej: __pycache__
Only in ./test: __pycache__
Only in .: .test2.py.swp
Only in .: .vscode
```
I'm happy with that.

**Comment by gselzer** [comment]
@ctrueden wrote:
>CTR FIXME - Include the following hint for better debugging:
>```
>jgo.jgo._logger.addHandler(logging.StreamHandler(sys.stderr))
>jgo.jgo._logger.setLevel(logging.DEBUG)
>scyjava.start_jvm()
>```
>And clean up this "Common Errors" subsection in general to be

I created a `Debugging Tips` section in `doc/Troubleshooting.md`, but let's solve this with a PR dedicated to https://github.com/imagej/pyimagej/issues/188. There's just too much going on in this PR and by creating a new issue we can get this merged ASAP :smile: 
more helpful and less scary.

**Comment by ctrueden** [comment]
> I don't think we should be excluding anything more.

The following things are erroneously (IMHO) included:

* `*.py` from the base directory, if you created any such files for local testing.
* `doc/rtd/_build/**`, which exists if you've run `make html` in the `doc/rtd` folder.
* `doc/.ipynb_checkpoints/**`, which exists if you've ever opened up any of the Jupyter notebooks.
* `doc/sample-data/**`, and `doc/**/*.ipynb`, which are not typically present in Python project source distributions and bloat the size of the archive.
* Furthermore, I would not include `doc/README.md` since it references the Jupyter notebooks, nor any `doc/*` subfolders (`cellprofiler`, `examples`, `rtd`)

I propose correcting these issues via the following `MANIFEST.in` file:
```
include LICENSE.txt
include MANIFEST.in
include README.md
include codecov.yml
include conftest.py
include dev-environment.yml
include environment.yml
include pyproject.toml
include setup.cfg
include setup.py
include test.sh
include doc/*.md
exclude doc/**/README.md
recursive-include src *.py
recursive-include test *.py
```

Which produces the following contents on my thoroughly messy working copy:
```
pyimagej-1.1.2.dev0/
pyimagej-1.1.2.dev0/LICENSE.txt
pyimagej-1.1.2.dev0/MANIFEST.in
pyimagej-1.1.2.dev0/PKG-INFO
pyimagej-1.1.2.dev0/README.md
pyimagej-1.1.2.dev0/codecov.yml
pyimagej-1.1.2.dev0/conftest.py
pyimagej-1.1.2.dev0/dev-environment.yml
pyimagej-1.1.2.dev0/doc/
pyimagej-1.1.2.dev0/doc/Development.md
pyimagej-1.1.2.dev0/doc/Headless.md
pyimagej-1.1.2.dev0/doc/Initialization.md
pyimagej-1.1.2.dev0/doc/Install.md
pyimagej-1.1.2.dev0/doc/Troubleshooting.md
pyimagej-1.1.2.dev0/doc/Xvfb.md
pyimagej-1.1.2.dev0/environment.yml
pyimagej-1.1.2.dev0/pyproject.toml
pyimagej-1.1.2.dev0/setup.cfg
pyimagej-1.1.2.dev0/setup.py
pyimagej-1.1.2.dev0/src/
pyimagej-1.1.2.dev0/src/imagej/
pyimagej-1.1.2.dev0/src/imagej/__init__.py
pyimagej-1.1.2.dev0/src/imagej/config.py
pyimagej-1.1.2.dev0/src/imagej/dims.py
pyimagej-1.1.2.dev0/src/imagej/stack.py
pyimagej-1.1.2.dev0/src/pyimagej.egg-info/
pyimagej-1.1.2.dev0/src/pyimagej.egg-info/PKG-INFO
pyimagej-1.1.2.dev0/src/pyimagej.egg-info/SOURCES.txt
pyimagej-1.1.2.dev0/src/pyimagej.egg-info/dependency_links.txt
pyimagej-1.1.2.dev0/src/pyimagej.egg-info/entry_points.txt
pyimagej-1.1.2.dev0/src/pyimagej.egg-info/requires.txt
pyimagej-1.1.2.dev0/src/pyimagej.egg-info/top_level.txt
pyimagej-1.1.2.dev0/test/
pyimagej-1.1.2.dev0/test/test_imagej.py
pyimagej-1.1.2.dev0/test.sh
```

**Comment by gselzer** [comment]
> > I don't think we should be excluding anything more.
> 
> The following things are erroneously (IMHO) included:
> 
>     * `*.py` from the base directory, if you created any such files for local testing.
> 
>     * `doc/rtd/_build/**`, which exists if you've run `make html` in the `doc/rtd` folder.
> 
>     * `doc/.ipynb_checkpoints/**`, which exists if you've ever opened up any of the Jupyter notebooks.
> 
>     * `doc/sample-data/**`, and `doc/**/*.ipynb`, which are not typically present in Python project source distributions and bloat the size of the archive.
> 
>     * Furthermore, I would not include `doc/README.md` since it references the Jupyter notebooks, nor any `doc/*` subfolders (`cellprofiler`, `examples`, `rtd`)
> 
> 
> I propose correcting these issues via the following `MANIFEST.in` file:

Good point, I hadn't run these things. We now have a much more declarative `MANIFEST.in`.

The only thing I would suggest we remove from your list is `codecov.yml`. I don't see why someone would want that.
Let me know what you think.

**Comment by ctrueden** [comment]
@gselzer In the interest of getting this excellent work onto the mainline sooner, I have merged. However, I did notice one additional small thing in testing the latest work: I still have `codecov.yml` in the sdist, even though it is not explicitly enumerated in the MANIFEST.in. That makes me a little nervous that other stuff could inadvertently slip in later. But it's fine enough for now!

**Pull request merged**